_posts/2017-11-21-Installing-Jenkins-with-PowerShell-DSC.md

```diff
@@ -0,0 +1,745 @@
+
+```powershell
+Install-Module cChoco -f
+Install-Module xNetworking -f
+Install-Module xWebAdministration -f
+```
+
+Three modules that provide you three type of functionality. Something to get stuff for Choco, configuring network stuff and managing IIS. If you run this for the first time you might get question about if you want to install nuget package provider for PowerShell. You should if you want to follow this path. It grabs you the wanted modules from [PowerShellGallery](https://www.powershellgallery.com/packages/cChoco/2.3.1.0). 
+
+## Our objective 
+
+We want to have a Jenkins server running in the end with following requirements:
+
+* Install all the needed dependencies
+* Install Jenkins itself
+* Configure Jenkins startup parameters 
+* Setup authentication for Jenkins 
+* Install plugins for Jenkins
+* Protect Jenkins by settings IIS in front of it 
+* Make sure that all traffic is HTTPS
+
+Once our objective is now clear we can look how I managed to do it.
+
+## Starting the scripting
+
+In the below I have the basic structure of my PowerShell script. There are few important parts in the script. The most important one is the Configuration JENKINS_CI which states that here is my DSC configuration. It has few steps in it:
+
+1. State the name of the configuration (JENKINS_CI)
+2. Declare parameters (JenkinsPort with default value of 8080)
+3. Import necessary DscResources (cChoco)
+4. For all the nodes make sure of the following
+5. Make sure NetFrameworkCore WindowsFeature is present
+6. Make sure Choco is installed
+
+```powershell
+Configuration JENKINS_CI
+{
+    param (
+        $JenkinsPort = 8080
+	)
+	
+	Import-DscResource -ModuleName 'cChoco'
+	
+	Node $AllNodes.NodeName {       
+		# Install .NET 3.5
+		WindowsFeature NetFrameworkCore 
+		{
+			Ensure    = "Present" 
+			Name      = "NET-Framework-Core"
+		}
+
+		# Install Chocolatey
+		cChocoInstaller installChoco
+		{
+			InstallDir = "c:\choco"
+			DependsOn = "[WindowsFeature]NetFrameworkCore"
+		}
+	}
+}
+```
+
+You should notice from above that you can mark one resource to be dependent on another with DependsOn. This is super important since we have multiple requirements for actually installing most of the stuff in our configuration later on. Now when we have the basic DSC configuration in place we can call it. 
+
+1. We need to define the AllNodes.NodeName that we referred in the script. 
+2. We also need to call give the parameters we want to the configuration. 
```

Comment:
- [ ] Tarkoititko: "need to **call** give the parameters" -> "need to give the parameters"?